cmd/kube-controller-manager/app/core.go

```diff
@@ -279,42 +277,33 @@ func startGarbageCollectorController(ctx ControllerContext) (bool, error) {
 		return false, nil
 	}
 
-	// TODO: should use a dynamic RESTMapper built from the discovery results.
-	restMapper := api.Registry.RESTMapper()
-
 	gcClientset := ctx.ClientBuilder.ClientOrDie("generic-garbage-collector")
-	preferredResources, err := gcClientset.Discovery().ServerPreferredResources()
-	if err != nil {
-		return true, fmt.Errorf("failed to get supported resources from server: %v", err)
-	}
-	deletableResources := discovery.FilteredBy(discovery.SupportsAllVerbs{Verbs: []string{"get", "list", "watch", "patch", "update", "delete"}}, preferredResources)
-	deletableGroupVersionResources, err := discovery.GroupVersionResources(deletableResources)
-	if err != nil {
-		return true, fmt.Errorf("Failed to parse resources from server: %v", err)
-	}
+
+	restMapper, syncThread := gcsetup.RESTMapper(gcClientset.Discovery(), 30*time.Second)
 
 	config := ctx.ClientBuilder.ConfigOrDie("generic-garbage-collector")
-	config.ContentConfig.NegotiatedSerializer = serializer.DirectCodecFactory{CodecFactory: metaonly.NewMetadataCodecFactory()}
-	metaOnlyClientPool := dynamic.NewClientPool(config, restMapper, dynamic.LegacyAPIPathResolverFunc)
 	config.ContentConfig = dynamic.ContentConfig()
+	// TODO: Make NewMetadataCodecFactory support arbitrary (non-compiled)
+	// resource types. Otherwise we'll be storing full Unstructured data in our
+	// caches for custom resources.
+	metaOnlyClientPool := dynamic.NewClientPool(config, restMapper, dynamic.LegacyAPIPathResolverFunc)
 	clientPool := dynamic.NewClientPool(config, restMapper, dynamic.LegacyAPIPathResolverFunc)
-
+	deletableResources, err := gcsetup.GetDeletableResources(gcClientset.Discovery())
+	if err != nil {
+		return true, err
+	}
 	ignoredResources := make(map[schema.GroupResource]struct{})
 	for _, r := range ctx.Options.GCIgnoredResources {
 		ignoredResources[schema.GroupResource{Group: r.Group, Resource: r.Resource}] = struct{}{}
 	}
-
-	garbageCollector, err := garbagecollector.NewGarbageCollector(
-		metaOnlyClientPool,
-		clientPool,
-		restMapper,
-		deletableGroupVersionResources,
-		ignoredResources,
-		ctx.InformerFactory,
-	)
+	garbageCollector, err := garbagecollector.NewGarbageCollector(metaOnlyClientPool, clientPool, restMapper, deletableResources, ignoredResources, ctx.InformerFactory)
 	if err != nil {
 		return true, fmt.Errorf("Failed to start the generic garbage collector: %v", err)
 	}
+
+	// Refresh discovery information periodically. Start/stop monitors.
+	go syncThread(garbageCollector, ctx.Stop)
```

Comment:
> If you feel strongly we can continue the discussion about a different factoring.

I feel strongly.  I'm not certain that the monitor sync has to line up exactly with the restmapper sync, but if it does, pass both through to the garbage collector and have the function live there.